Cost Estimation UI for Clusters

mrsheepuk opened the pull request:
## Summary

Implements UI for cluster cost estimation, working for both AWS and GCP.

**Which issue(s) this PR resolves**:
Resolves #961 

## Checklist

 - [x] I've created a PR to update the [documentation](https://github.com/appvia/kore-docs): appvia/kore-docs#18

## Changes

* UI cost estimation / display for clusters. ef343eed
  * Use the metadata from the API to enrich UI with region drop-downs and machine-type drop-downs for both GCP and AWS
  * Calculate 'on the fly' estimates for node pools as the user edits
  * Request full plan estimates from the API when creating, editing and using plans.
* Ensure API estimates costs of non-autoscale clusters. d903398
* Sort schemas for GKE and EKS logically instead of alphabetically to improve UI display ordering. d32a8a2

### Additional changes

* Use a patched cloudinfo to work around issue retrieving AWS information in some regions. 5d1811c

## Screenshots

Cost estimate on plan create/edit:

![image](https://user-images.githubusercontent.com/7505593/87079080-74583780-c21d-11ea-9b5f-611077dffc12.png)

Cost estimate on node pool editor:

![image](https://user-images.githubusercontent.com/7505593/87079131-84701700-c21d-11ea-8d52-39e0b3da4afe.png)

## Examples

Enable costs support by applying `examples/kore-costs-feature.yaml` then play with the UI. You will need to ensure that the credential references in the kore-costs-feature point to valid AWS and GCP credentials.

## Testing

* Add new cluster plans for AWS and GCP
* Edit existing cluster plans for AWS and GCP
* Edit plan when creating a cluster for AWS and GCP and see the estimated costs being shown.

## Follow-up stories

* #1049 - use the new metadata API to power K8S version selection controls on the UI.
* #1050 - show cost estimates when viewing/editing existing clusters and viewing existing plans.

davet1985 commented on ui/lib/components/costs/CostBreakdown.js:
```diff
@@ -0,0 +1,43 @@
+        pagination={false} 
+        dataSource={rows} 
+        columns={columns} 
+        footer={() => <>Cost basis: <Switch className="double-on-switch" checkedChildren="Monthly" unCheckedChildren="Hourly" onChange={(c) => this.setState({ monthly: c })} defaultChecked={true} /></>} 
```
What do you think about this being a solid style radio button, I think it would be more user friendly both options will be displayed to the user and switches are naturally on/off

mrsheepuk replied:
Addressed in b06155e

davet1985 commented on ui/lib/components/costs/CostEstimate.js:
```diff
@@ -0,0 +1,66 @@
+
+  render() {
+    const { estimate, planValuesChangedSinceEstimate, estimateError } = this.state
+    if (!estimate && !estimateError && this.props.estimateInit && this.props.planValues) {
+      this.estimate()
+    }
```
Could this logic be handled in `componentDidMount` or `componentDidUpdate`? Something about it being in the render doesn't feel right...

mrsheepuk replied:
you're quite right, it was late... let me fix that up.

mrsheepuk replied:
Fixed b06155e

davet1985 commented on ui/lib/components/costs/NodePoolCost.js:
```diff
@@ -0,0 +1,52 @@
+
+    const nodePoolCosts = { minCost: null, maxCost: null, typicalCost: null }
+    const zoneMultiplier = this.props.zoneMultiplier ? this.props.zoneMultiplier : 1
+    const size = nodePool.size || nodePool.desiredSize // annoying diff between GKE and EKS naming
```
Like you say we should amend the naming to match as this is annoying and potentially fragile

mrsheepuk replied:
yes... the AKS is going to follow the GKE naming convention so at some point we'll need to take the hit and refactor the EKS node groups to match those.... not looking forward to that.

davet1985 commented on ui/lib/components/plans/PlanOptionBase.js:
```diff
@@ -19,6 +19,7 @@ export default class PlanOptionBase extends React.Component {
     mode: PropTypes.oneOf(['create','view','edit']),
     team: PropTypes.object, // may be optionally used by custom plan option components to give richer interface when manage=false.
     id: PropTypes.string,
+    disableCustom: PropTypes.bool, // set to true to force not using a custom control
```
What scenarios do we want to avoid using custom controls? (just curious)

mrsheepuk replied:
there's a specific use case already for this - basically, in a custom control itself, if it decided it can't work, it can re-call planoption with disableCustom true - for an example, see the PlanOptionClusterMachineType line 194.

davet1985 commented on ui/lib/components/plans/custom/PlanOptionClusterRegion.js:
```diff
@@ -0,0 +1,84 @@
+    this.componentDidMountComplete = Promise.resolve().then(async() => {
+      this.setState({ loadingRegions: true })
+      const provider = this.props.kind
+      const api = await KoreApi.client()
```
Should probably put this inside the try in case there's an error getting the client?

mrsheepuk replied:
Fixed b06155e

davet1985 commented on ui/lib/components/plans/custom/PlanOptionEKSNodeGroups.js:
```diff
@@ -161,16 +140,16 @@ export default class PlanOptionEKSNodeGroups extends PlanOptionBase {
     if (selected) {
       amiType = selected.amiType || 'AL2_x86_64'
-      instanceTypes = PlanOptionEKSNodeGroups.supportedInstanceTypes[amiType]
+      // instanceTypes = PlanOptionEKSNodeGroups.supportedInstanceTypes[amiType]
```
Should this still be commented?

mrsheepuk replied:
Removed b06155e

davet1985 commented on ui/lib/components/plans/custom/PlanOptionEKSNodeGroups.js:
```diff
@@ -218,7 +197,7 @@ export default class PlanOptionEKSNodeGroups extends PlanOptionBase {
                         {this.validationErrors(`${name}[${selectedIndex}].minSize`)}
                       </Descriptions.Item>}
-                      <Descriptions.Item label={selected.enableAutoscaler ? 'Initial' : null}>
+                      <Descriptions.Item label={selected.enableAutoscaler ? 'Initial Size' : null}>
```
Lower case "size"?

davet1985 commented on ui/lib/components/plans/custom/PlanOptionGKENodePools.js:
```diff
@@ -325,7 +204,7 @@ export default class PlanOptionGKENodePools extends PlanOptionBase {
                         {this.validationErrors(`${name}[${selectedIndex}].minSize`)}
                       </Descriptions.Item>}
-                      <Descriptions.Item label={selected.enableAutoscaler ? 'Initial' : null}>
+                      <Descriptions.Item label={selected.enableAutoscaler ? 'Initial Size' : null}>
```
Lower case "size"?

davet1985 commented:
I haven’t applied the korefeature CRD yet but I’m getting costs stuff appearing on the UI

davet1985 commented:
When I configured the KoreFeature locally, the Kore secret wasn't set up as expected which led to an Error in the feature. I edited the Kore secret to fix the error, but the KoreFeature did not pick up the change and fix itself. I had to remove the KoreFeature and re-apply it.

davet1985 commented:
I've tested locally and it seems to be working as expected, apart from the two comments above.

mrsheepuk commented:
@davet1985 addressed your comments in b06155e - going to look at the use case for costs-not-enabled now and make sure everything works well even if costs is not set up.

mrsheepuk commented:
@davet1985 This is ready for a final review now - added two further commits which are worth a look over, now tested and working with and without the kore-costs feature applied; there are now hard-coded fallbacks for node types and regions so the UI still gets some of the benefits if cloudinfo isn't available.

davet1985 commented on ui/lib/components/plans/PlanViewer.js:
```diff
@@ -1,22 +1,24 @@
     resourceType: PropTypes.oneOf(['cluster', 'service']).isRequired,
-    displayUnassociatedPlanWarning: PropTypes.bool
+    displayUnassociatedPlanWarning: PropTypes.bool,
+    hideCostEstimate: PropTypes.bool
```
I might be missing it, but I can't see any usage of this prop?

mrsheepuk replied:
line 35 :)

davet1985 replied:
Ha yeah, I saw those ones. I meant being passed to the component as otherwise costs would always be hidden?

mrsheepuk replied:
Oh I see what you mean - yeah - it's for the future use case of being able to choose whether to show cost estimates to team members or not, not actually used yet, but will be.

mrsheepuk commented on ui/lib/components/costs/CostEstimate.js:
```diff
@@ -0,0 +1,70 @@
+        <CostBreakdown costs={estimate} style={{ marginBottom: '20px' }} />
+        {!estimateError ? null : (
+          <>
+            <Alert type="warn" message={`Unable to prepare estimate - ${estimateError.message}`} />
```
typo: this should be 'warning'

davet1985 commented:
I've re-tested the branch and it's working as expected. Review the UI code and LGTM

bandesz commented:
I've created an issue to validate the KoreFeature parameters, as you can set non-existing parameters without any errors: https://github.com/appvia/kore/issues/1064

bandesz commented:
When I open the prepare "Cluster running cost estimate" I shouldn't need to push a button, but we should trigger the cost estimation simply by opening the panel

bandesz commented:
"Plan values have changed - click to refresh" - Not everyone will realise this button text and seems like an extra click we should automate. IMO when the plan edit drawer is closed, we should trigger an estimation refresh.

bandesz commented:
Note to my previous commit: I think in every case we should automatically refresh the estimate if it's visible and there is no active drawer. (E.g. when I change the cluster plan)

bandesz commented:
When creating a plan you can estimate costs without adding any pools. It's a small thing, but maybe we shouldn't estimate until a region is selected and at least one nodepool is added. TBH having a default nodepool set in the form would be even better and solve this issue.

bandesz commented on pkg/costs/metadata.go:
```diff
@@ -45,17 +46,39 @@ type Metadata interface {
+
+func (m *metadataImpl) info() Cloudinfo {
+	// Check every 30 seconds if cloudinfo is available.
+	if m.cloudinfoCheckDue.IsZero() || m.cloudinfoCheckDue.Before(time.Now()) {
```
Just a note that this method can be called multiple times in parallel, but it seems thread-safe. The worst that can happen is we call the API twice at the same time.

mrsheepuk replied:
good point - but yes, I think synchronising it would be an overkill as you rightly point out the worst case impact is two calls to the status endpoint of cloudinfo.

bandesz commented on pkg/serviceproviders/application/helper.go:
```diff
@@ -116,13 +117,32 @@ func compileResource(obj runtime.Object, params ResourceParams) (runtime.Object,
 		New("document").
 		Funcs(template.FuncMap{
 			"json": func(v interface{}) interface{} {
-				val, _ := json.Marshal(v)
```
The template functions can return an error as a second argument, so we shouldn't swallow them.

mrsheepuk replied:
done, see 2beef6c

bandesz commented:
One more usability question: maybe we should check whether the kore-costs feature is enabled everywhere, so we can either hide the functionality or show the user that the feature is not enabled (so they don't get an error message when clicking the button)

mrsheepuk commented:
Updates made:

* Removed 'prepare estimate' button, now auto-populate the estimate once enough data is available, and auto-refresh (with a 1-second debounce) whenever plan values change. Added a loading indicator during that 1s so you can see visually it's about to change.
* Validated node pool properties correctly on the back end, to prevent panics if you manage to submit a node pool for estimation with no name / size / etc.
* Validated at least one node pool present on the back end before producing an estimate, as per the region.
* Returned a specific validation error where pricing info is not available and put some custom messaging around this in the UI to drive the user to the right course of action to enable price estimation.
* No longer swallowing errors in the template functions.

All in 2beef6c3

mrsheepuk commented:
Fixed up the UI E2E tests (grrrrr) and all comments addressed - @bandesz if you're happy could you approve?
